tests/tests/test_update.py

```diff
@@ -0,0 +1,191 @@
+from common_docker import *
+from mendertesting import MenderTesting
+from common_update import common_update_procedure
+from MenderAPI import auth, adm, deploy, inv, deviceauth
+import subprocess
+import time
+import conftest
+
+
+def setup_docker_volumes():
+    docker_volumes = ["mender-artifacts",
+                      "mender-deployments-db",
+                      "mender-deviceadm-db",
+                      "mender-deviceauth-db",
+                      "mender-dynomite-db",
+                      "mender-elasticsearch-db",
+                      "mender-inventory-db",
+                      "mender-useradm-db"]
+
+    for volume in docker_volumes:
+        ret = subprocess.call(["docker", "volume", "create", "--name=%s" % volume])
+        assert ret == 0, "failed to create docker volumes"
+
+
+def setup_upgrade_source(upgrade_from):
+    # TODO: change this
+    if not os.path.exists("upgrade-from"):
+        ret = subprocess.call(["git", "clone", "https://github.com/GregorioDiStefano/integration", "-b", upgrade_from, "upgrade-from"])
+        assert ret == 0, "failed to git clone old source version"
+
+    # TODO: remove TEST_BRANCH: 1.0.x once merged
+    # start upgrade source and bring up it's production environment
+    ret = subprocess.call(["bash", "run.sh", "--get-requirements"],
+                          cwd="upgrade-from/tests",
+                          env={"TEST_BRANCH": "1.0.1"})
+
+    assert ret == 0, "failed running 'run.sh --get-requirements' on upgrade source"
+    ret = subprocess.call(["python", "production_test_env.py", "--start"],
+                          cwd="upgrade-from/tests")
+    assert ret == 0, "failed running 'production_test_env.py start' on upgrade source"
+
+
+def perform_upgrade():
+    ret = subprocess.call(["cp", "-r", "tests/upgrade-from/keys-generated", "."], cwd="..")
+    assert ret == 0, "faled to copy keys from original environment"
+
+    ret = subprocess.call(["cp", "extra/production-environment/production-testing-env.yml", "."], cwd="..")
+    assert ret == 0, "copying extra/production-envrionment folder failed."
+
+    ret = subprocess.call(["docker-compose", "-p", "testprod", "-f", "docker-compose.yml", "-f", "docker-compose.storage.minio.yml", "-f", "production-testing-env.yml", "up", "-d"], cwd="..")
+    assert ret == 0
+
+    # give time for all microservices to come online
+    time.sleep(30)
+
+
+def setup_fake_clients(device_count, fail_count):
+    p = subprocess.Popen(["mender-stress-test-client",
+                         "-count=%s" % str(device_count),
+                         "-invfreq=1",
+                         "-pollfreq=1",
+                         "-wait=1",
+                         "-failcount=%s" % str(fail_count)])
+    return p
+
+
+def provision_upgrade_server():
+    # deploy update to 10 devices
+    ret = subprocess.Popen(["python", "production_test_env.py", "--test-deployment"], cwd="upgrade-from/tests", stdout=subprocess.PIPE)
+    time.sleep(120)
```

Comment:
Can you replace this with a `ssh_is_opened` call to the client? Or something similar to it. The client is always the last to finish and this has worked really well for our other tests. This gives two great benefits: You both finish faster if you get a fast droplet, and will in fact wait longer (but not fail) if you get a crappy droplet.